File: FORTRAN/STENCIL/ms_cdr.f90
program ms_stencil
!
! This programme illustrates the use of the idrs-package on
! a convection-diffusion-reaction problem with constant coefficients
! and homogeneous dirichlet boundary conditions.
!
! The problem can be  solved with the different algorithms: IDRS, BiCGSTAB and QMRIDR,
! Neumann and Chebyshev polynomial preconditioners, and with and without recycling techniques.
! The problem is solved with the different idrs-algorithms: IDRS and QMRIDR 
! for the single shift (= reaction parameter) problem and MSIDRS and MSQMRIDR 
! for the multishift problem, which uses several reaction parameters.
!
! The matrix is not explicitly stored. The matrix-vector multiplication
! is done using a stencil operations.
!
! This software is distributed under the MIT License:
! http://www.opensource.org/licenses/mit-license.php
! Copyright:(c) 2024 Martin van Gijzen
!

   use interface_module          ! To read command line options
   use precision_module          ! Set real and complex precision
   use matrix_module             ! Always needed is a matrix module that defines the
                                 ! matrix-vector multiplication and preconditioning operation
   use user_module               ! User functions, user defined matrix-vector product
   use idrs_module               ! The module with the IDRS routines
   use ritz_module               ! Can be used for computing an initial search space 
                                 ! and user defined omegas

   implicit none 

   integer                       :: grid
   integer, parameter            :: default_grid = 3
!
   integer, parameter            :: Nsigma = 6              ! Number of sigma
   real(kind=rp)                 :: sigma(Nsigma)           ! Shifts
   real(kind=rp)                 :: dsigma                  ! Shift spacing 
   real(kind=rp)                 :: sigma_max               ! Maximum shift
   real, parameter               :: default_sigma_max = 5.  ! Maximum shift
!
   type(user_matrix)             :: A1, M1

   real(kind=rp), allocatable    :: x(:,:), x0(:), b(:)

   integer                       :: tb, te, clock_rate, clock_max  ! For system_clock
   integer                       :: my_proc, n_procs

   real(kind=rp)                 :: radius
   real(kind=rp)                 :: true_relres(Nsigma), normb
   integer                       :: i

! Declaration and initialization of IDRS parameters:
   include "../INCLUDE/initialize.inc"

   my_proc = this_image()
   n_procs = num_images()

! Which grid?
   grid = get_parameter( '-grid', default_grid )

! Make the matrices
   call cdr_matrix( A1, grid )
   call mass_matrix( A1, M1 )

! Output message, problem description:
   if ( this_image() == 1 ) then
      write(*,*)
      write(*,'(a)') '========================================================================='
      write(*,'(6x,a,i10,/)') &
         'Convection-diffusion-reaction problem, number of equations is ', n_procs*A1%Nn
   end if
!
! Output message, which methods:
   include "../INCLUDE/output_methods.inc"

! Allocate solution and right-hand-side vectors:
   allocate( x0(A1%Nn), b(A1%Nn), x(A1%Nn,Nsigma), resnrm(Nsigma) )

   A1%multishift = .true.

! Define the shifts:
   sigma_max = get_parameter( '-reaction', default_sigma_max )
   dsigma = 0.
   if ( Nsigma > 1 ) dsigma = sigma_max/(Nsigma-1)
   do i = 1, Nsigma
      sigma(i) = (i-1)*dsigma
   end do

! Compute right-hand-side
   call model_solution( x0, A1 )
   b = A1*x0
   normb = norm(b)

! Store the matrix in correct form
   A = real_matrix( A1 )

! Compute estimates for largest and smallest eigenvalues with Gershgorin's theorem:
   call gershgorin( M1, rcenter, radius )
   rfoci(1) = rcenter-radius
   rfoci(2) = rcenter+radius  
! Approximate inverse of Mass matrix by polynomial:
   call real_preconditioner( M, M1, preconditioner=preconditioner, degree=degree, &
         foci=rfoci, center=rcenter )

!
! Test the multishift algorithms
   call system_clock ( tb, clock_rate, clock_max )
   if ( use_idrs ) then
      variant = 1
      x = MSIDRS( A, b, sigma, s, M, tol, maxit, variant, flag, resnrm, iter )
   elseif ( use_bicgstab ) then
      variant = 2
      x = MSIDRS( A, b, sigma, s, M, tol, maxit, variant, flag, resnrm, iter )
   elseif ( use_qmridr ) then
      x = MSQMRIDR( A, b, sigma, s, M, tol, maxit, flag, resnrm, iter, in_s, in_tol, in_it )
   else
      error stop 'Error: no solution method specified!'
   end if
   call system_clock ( te, clock_rate, clock_max )

!
! Compute true residual norms
   do i = 1, Nsigma
      true_relres(i) = norm( b - A1*x(:,i) + sigma(i)*(M1*x(:,i)) )/normb
   end do

   if ( my_proc == 1 ) then
      write(*,'(a)') 'Results: '
      write(*,'(a,f8.2)') 'Elapsed time         = ', real ( te - tb ) / real ( clock_rate )
      write(*,'(a,i4)')   'Number of iterations = ', iter
      do i = 1, Nsigma
         write(*,'(a,f8.2,2x,a,e10.2)') 'Shift = ', sigma(i), '||r||/||b|| = ', resnrm(i)
         write(*,'(a,e10.2)') 'True relative residual norm   = ', true_relres(i)
      end do
      if ( flag > 0 ) then
         if ( flag == 1 ) write(*,*) 'Maximum number of iterations reached!'
         if ( flag == 2 ) write(*,*) 'Accuracy above prescribed tolerance!'
         if ( flag == 3 ) write(*,*) 'Break down!'
      end if
      write(*,'(a)') &
         '==============================================================================='
   end if

end program ms_stencil

File: MATLAB/STENCIL/ms_cdr.m
%
% IDRS parameterised testproblems
%
% The software is distributed without any warranty.
%
% Martin van Gijzen
% Copyright (c) December 2023
%

clear all;
close all;
clc;

disp('FDM discretisation of a 3D convection-diffusion-reaction problem on a unit cube');
disp('+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++');

% Define system

% Defaults:
m = 64;
eps = 0.005;
beta(1) = 1;
beta(2) = 1;
beta(3) = 1;

% Defaults for the iterative solvers:

tol = 1e-6;
maxit = 1000;

Title = 'Parameters';
refine = 1;
variant = 'idrs';
s = 4;
tol = 1e-6;
maxit = 10000;
in_s = 4;
in_tol = 1e-1;
in_it = 0;

contin = 1;
while contin
%% Menu
   prompt = { 'Refinement in each direction:', ...
              'Method: idrs/bicgstab/qmridr',...
              'Parameter s:', ...
              'Tolerance:',...
              'Maximum iterations ',...
              'Inner s:',...
              'Inner tolerance:',...
              'Inner iterations:'};
   defaults = { num2str(refine), variant, num2str(s), num2str(tol), num2str(maxit), num2str(in_s), num2str(in_tol), num2str(in_it)};
   lineNo=[1 50];
   params=inputdlg(prompt,Title,lineNo,defaults);
   if ( isempty(params) ) contin = 0; break; end;
   refine   = str2num(char(params(1)));
   variant  =         char(params(2));
   s        = str2num(char(params(3)));
   tol      = str2num(char(params(4)));
   maxit    = str2num(char(params(5)));
   in_s     = str2num(char(params(6)));
   in_tol   = str2num(char(params(7)));
   in_it    = str2num(char(params(8)));

   method = 0;
   if ( strcmpi(variant,'idrs' ) )
      method = 1;
   elseif ( strcmpi(variant,'bicgstab' ) )
      method = 2;
      s = 1;
   end

% Generate matrix
   m = 64*refine;
   h = 1/(m+1);

   n = m*m*m;
   Sx = gallery('tridiag',m,-eps/h^2-beta(1)/(2*h),2*eps/h^2,-eps/h^2+beta(1)/(2*h));
   Sy = gallery('tridiag',m,-eps/h^2-beta(2)/(2*h),2*eps/h^2,-eps/h^2+beta(2)/(2*h));
   Sz = gallery('tridiag',m,-eps/h^2-beta(3)/(2*h),2*eps/h^2,-eps/h^2+beta(3)/(2*h));
   Is = speye(m,m);
   A = kron(kron(Is,Is),Sx) + kron(kron(Is,Sy),Is)+ kron(kron(Sz,Is),Is);

   x = linspace(h,1-h,m);
   sol = kron(kron(x.*(1-x),x.*(1-x)),x.*(1-x))';
   sol = sqrt(sol);
   b = A*sol;

   n_shift = 6;
   shift = [0, 1, 2, 3, 4, 5];

%
%    Output: description of experiment:
%
%    Initialize figures
   scrsz = get(0,'ScreenSize');
   fig1 = figure('Position',[scrsz(1) 0.5*scrsz(4) 0.4*scrsz(3) 0.5*scrsz(4)]);
   if ( strcmpi(variant,'bicgstab') )
      title(['Convergence msbicgstab, ']);
   else
      title(['Convergence ', variant, s]);
   end
   xlabel('Number of iterations');
   ylabel('Scaled residual norm');
   grid on;
   hold on;
%
   t = cputime;
   if ( method == 1 || method == 2 )
      [x,flag,relres,iter,resvec] = ...
          msidrs(A,b,shift,s,[],tol,maxit,method);
   else
      [x,flag,relres,iter,resvec] = ...
          msqmridr(A,b,shift,s,[],tol,maxit,in_s,in_tol,in_it);
   end

   time = cputime - t;
   disp(['Multi-shift test problem, size is ',num2str(n)]);
   disp(['Solution method is ', variant, ' with s = ', num2str(s)]);
   disp(['Elapsed time            = ',num2str(time),'s.']);
   disp(['Number of iteration     = ',num2str(iter)]);
   for i = 1:n_shift
      disp(['Relative residual norm for shift ', num2str(shift(i)), ' = ', num2str(relres(i))]);
   end
   if ( flag > 0 )
      if ( flag == 1 ) disp('Maximum number of iterations reached!'); end;
      if ( flag == 2 ) disp('Accuracy above prescribed tolerance!'); end;
      if ( flag == 3 ) disp('Break down!'); end;
   end
%
   disp('==================================================');

%
% Plot convergence
   figure(fig1);
   for i_shift = 1:n_shift
      x_as = [0:1:iter];
      plot(x_as,log10(resvec(:,i_shift)/resvec(1,i_shift)));
%     legend(['shift=',num2str(shift(i_shift))]);
   end

end
